File: lib/domain/cultivation_plan/mappers/plan_allocation_adjust_read_snapshot_parts.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Mappers
      # PlanAllocationAdjustReadSnapshot 組み立ての純粋部分（I/O なし）。
      module PlanAllocationAdjustReadSnapshotParts
        module_function

        # @param stored_days [Integer, nil]
        # @param start_date [Date]
        # @param completion_date [Date]
        # @return [Integer]
        def effective_cultivation_days(stored_days:, start_date:, completion_date:)
          return stored_days if stored_days

          (completion_date - start_date).to_i + 1
        end

        # @param crop_stage_count [Integer]
        # @return [Boolean]
        def has_growth_stages?(crop_stage_count:)
          crop_stage_count.to_i.positive?
        end

        # @param estimated_cost [Numeric, nil]
        # @return [Float]
        def effective_estimated_cost(estimated_cost:)
          (estimated_cost || 0.0).to_f
        end

        # @param plan_field_snapshots [Array<#id,#name,#area,#daily_fixed_cost>]
        # @param field_cultivation_snapshots [Array<Dtos::PlanAllocationAdjustFieldCultivationSnapshot>]
        # @return [Array<Dtos::PlanAllocationAdjustFieldSourceSnapshot>]
        def build_field_source_snapshots(plan_field_snapshots:, field_cultivation_snapshots:)
          cultivations_by_field = field_cultivation_snapshots.group_by(&:field_id)

          plan_field_snapshots.map do |field|
            cultivations = cultivations_by_field[field.id] || []
            allocation_snapshots = cultivations.map { |fc| field_cultivation_allocation_snapshot(fc) }

            Dtos::PlanAllocationAdjustFieldSourceSnapshot.new(
              field_id: field.id,
              field_name: field.name,
              field_area: field.area,
              cultivations: allocation_snapshots
            )
          end
        end

        # @param crop_id [Integer]
        # @param crop_name [String]
        # @param groups [Object]
        # @param crop_stage_count [Integer]
        # @param build_agrr_requirement [Proc, nil] 生育ステージありのときだけ呼ばれる
        # @return [Dtos::PlanAllocationAdjustReadSnapshot::PlanCropSnapshot]
        def plan_crop_snapshot(crop_id:, crop_name:, groups:, crop_stage_count:, build_agrr_requirement: nil)
          has_growth = has_growth_stages?(crop_stage_count: crop_stage_count)
          requirement = has_growth && build_agrr_requirement ? build_agrr_requirement.call : nil

          Dtos::PlanAllocationAdjustReadSnapshot::PlanCropSnapshot.new(
            crop_id: crop_id,
            crop_name: crop_name,
            groups: groups,
            has_growth_stages: has_growth,
            agrr_requirement: requirement
          )
        end

        # @param field_cultivation [Dtos::PlanAllocationAdjustFieldCultivationSnapshot]
        def field_cultivation_allocation_snapshot(field_cultivation)
          Dtos::PlanAllocationAdjustFieldCultivationAllocationSnapshot.new(
            field_cultivation_id: field_cultivation.field_cultivation_id,
            field_id: field_cultivation.field_id,
            crop_id: field_cultivation.crop_id,
            crop_name: field_cultivation.crop_name,
            variety: field_cultivation.variety,
            area: field_cultivation.area,
            start_date: field_cultivation.start_date,
            completion_date: field_cultivation.completion_date,
            cultivation_days: effective_cultivation_days(
              stored_days: field_cultivation.stored_cultivation_days,
              start_date: field_cultivation.start_date,
              completion_date: field_cultivation.completion_date
            ),
            estimated_cost: effective_estimated_cost(estimated_cost: field_cultivation.estimated_cost),
            revenue: revenue_from_optimization_result(field_cultivation.optimization_result),
            accumulated_gdd: accumulated_gdd_from_optimization_result(field_cultivation.optimization_result),
            has_growth_stages: has_growth_stages?(crop_stage_count: field_cultivation.crop_stage_count)
          )
        end
        private_class_method :field_cultivation_allocation_snapshot

        def revenue_from_optimization_result(opt)
          return 0.0 unless opt.is_a?(Hash)

          (opt["revenue"] || opt["expected_revenue"] || 0.0).to_f
        end

        def accumulated_gdd_from_optimization_result(opt)
          return 0.0 unless opt.is_a?(Hash)

          value = opt["accumulated_gdd"]
          value = opt.dig("raw", "total_gdd") if value.nil?
          value.to_f
        end

        # @param weather_location [Domain::WeatherData::Contracts::WeatherLocationPredictionInput, nil]
        def weather_location_facts(weather_location)
          return {} if weather_location.nil?

          {
            latitude: weather_location.latitude.to_f,
            longitude: weather_location.longitude.to_f,
            elevation: (weather_location.elevation || 0.0).to_f,
            timezone: weather_location.timezone
          }
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/mappers/plan_allocation_adjust_read_snapshot_parts.rs
//! Ruby: `Domain::CultivationPlan::Mappers::PlanAllocationAdjustReadSnapshotParts`

use std::collections::BTreeMap;

use serde_json::{json, Value};

use crate::cultivation_plan::dtos::{
    PlanAllocationAdjustFieldCultivationAllocationSnapshot,
    PlanAllocationAdjustFieldCultivationSnapshot, PlanAllocationAdjustFieldSourceSnapshot,
    PlanAllocationAdjustPlanCropSnapshot, PlanAllocationAdjustPlanFieldSnapshot,
};
use crate::weather_data::dtos::WeatherLocation;

pub struct PlanAllocationAdjustReadSnapshotParts;

impl PlanAllocationAdjustReadSnapshotParts {
    pub fn effective_cultivation_days(
        stored_days: Option<i32>,
        start_date: time::Date,
        completion_date: time::Date,
    ) -> i32 {
        if let Some(days) = stored_days {
            return days;
        }
        (completion_date - start_date).whole_days() as i32 + 1
    }

    pub fn has_growth_stages(crop_stage_count: i32) -> bool {
        crop_stage_count > 0
    }

    fn effective_estimated_cost(estimated_cost: Option<f64>) -> f64 {
        estimated_cost.unwrap_or(0.0)
    }

    pub fn build_field_source_snapshots(
        plan_field_snapshots: &[PlanAllocationAdjustPlanFieldSnapshot],
        field_cultivation_snapshots: &[PlanAllocationAdjustFieldCultivationSnapshot],
    ) -> Vec<PlanAllocationAdjustFieldSourceSnapshot> {
        let mut cultivations_by_field: BTreeMap<i64, Vec<&PlanAllocationAdjustFieldCultivationSnapshot>> =
            BTreeMap::new();
        for fc in field_cultivation_snapshots {
            cultivations_by_field
                .entry(fc.field_id)
                .or_default()
                .push(fc);
        }

        plan_field_snapshots
            .iter()
            .map(|field| {
                let cultivations = cultivations_by_field
                    .get(&field.id)
                    .map(|rows| {
                        rows.iter()
                            .map(|fc| Self::field_cultivation_allocation_snapshot(fc))
                            .collect()
                    })
                    .unwrap_or_default();

                PlanAllocationAdjustFieldSourceSnapshot::new(
                    field.id,
                    field.name.clone(),
                    field.area,
                    cultivations,
                )
            })
            .collect()
    }

    pub fn plan_crop_snapshot<F>(
        crop_id: i64,
        crop_name: impl Into<String>,
        groups: Value,
        crop_stage_count: i32,
        build_agrr_requirement: Option<F>,
    ) -> PlanAllocationAdjustPlanCropSnapshot
    where
        F: FnOnce() -> Value,
    {
        let has_growth = Self::has_growth_stages(crop_stage_count);
        let requirement = if has_growth {
            build_agrr_requirement.map(|f| f())
        } else {
            None
        };

        PlanAllocationAdjustPlanCropSnapshot::new(
            crop_id,
            crop_name,
            groups,
            has_growth,
            requirement,
        )
    }

    pub fn weather_location_facts(
        weather_location: &WeatherLocation,
    ) -> BTreeMap<String, Value> {
        BTreeMap::from([
            ("latitude".into(), json!(weather_location.latitude)),
            ("longitude".into(), json!(weather_location.longitude)),
            (
                "elevation".into(),
                json!(weather_location.elevation.unwrap_or(0.0)),
            ),
            (
                "timezone".into(),
                json!(weather_location.timezone.clone().unwrap_or_default()),
            ),
        ])
    }

    fn field_cultivation_allocation_snapshot(
        field_cultivation: &PlanAllocationAdjustFieldCultivationSnapshot,
    ) -> PlanAllocationAdjustFieldCultivationAllocationSnapshot {
        PlanAllocationAdjustFieldCultivationAllocationSnapshot::new(
            field_cultivation.field_cultivation_id,
            field_cultivation.field_id,
            field_cultivation.crop_id.to_string(),
            field_cultivation.crop_name.clone(),
            field_cultivation.variety.clone(),
            field_cultivation.area,
            field_cultivation.start_date,
            field_cultivation.completion_date,
            Self::effective_cultivation_days(
                field_cultivation.stored_cultivation_days,
                field_cultivation.start_date,
                field_cultivation.completion_date,
            ),
            Self::effective_estimated_cost(field_cultivation.estimated_cost),
            revenue_from_optimization_result(field_cultivation.optimization_result.as_ref()),
            accumulated_gdd_from_optimization_result(
                field_cultivation.optimization_result.as_ref(),
            ),
            Self::has_growth_stages(field_cultivation.crop_stage_count),
        )
    }
}

fn revenue_from_optimization_result(opt: Option<&Value>) -> f64 {
    let Some(Value::Object(map)) = opt else {
        return 0.0;
    };
    map.get("revenue")
        .or_else(|| map.get("expected_revenue"))
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0)
}

fn accumulated_gdd_from_optimization_result(opt: Option<&Value>) -> f64 {
    let Some(Value::Object(map)) = opt else {
        return 0.0;
    };
    if let Some(v) = map.get("accumulated_gdd").and_then(|v| v.as_f64()) {
        return v;
    }
    map.get("raw")
        .and_then(|raw| raw.get("total_gdd"))
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0)
}

#[cfg(test)]
mod mappers_plan_allocation_adjust_read_snapshot_parts_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/cultivation_plan/mappers_plan_allocation_adjust_read_snapshot_parts_test.rs"));
}
